src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Lower.java

```diff
@@ -1794,18 +1797,30 @@ JCStatement initField(int pos, Symbol rhs, Symbol lhs) {
                     make.Ident(rhs)).setType(lhs.erasure(types)));
     }
 
-    /** Return tree simulating the assignment {@code this.this$n = this$n}.
+    /**
+     * Return tree simulating null checking outer this and/or assigning. This is
+     * called when a null check is required (nullCheckOuterThis), or a synthetic
+     * field is generated (stores).
      */
-    JCStatement initOuterThis(int pos, VarSymbol rhs) {
+    JCStatement initOuterThis(int pos, VarSymbol rhs, boolean stores) {
         Assert.check(rhs.owner.kind == MTH);
-        VarSymbol lhs = outerThisStack.head;
-        Assert.check(rhs.owner.owner == lhs.owner);
         make.at(pos);
-        return
-            make.Exec(
-                make.Assign(
+        JCExpression expression;
+        if (stores) {
+            VarSymbol lhs = outerThisStack.head;
+            Assert.check(rhs.owner.owner == lhs.owner);
+            JCExpression sourceExp = make.Ident(rhs);
+            if (nullCheckOuterThis) {
+                sourceExp = attr.makeNullCheck(sourceExp);
+            }
+            expression = make.Assign(
                     make.Select(make.This(lhs.owner.erasure(types)), lhs),
-                    make.Ident(rhs)).setType(lhs.erasure(types)));
+                    sourceExp).setType(lhs.erasure(types));
+        } else {
+            Assert.check(nullCheckOuterThis);
```

Comment:
This code seems repeated twice. Perhaps you could first create the expression (which could be null checked or not) and then use that expression to store in the `this$0` field (if needed).

Reply:
Thanks for the comment. Done.